src/nerdbank-streams/src/Utilities.ts

```diff
@@ -72,32 +72,83 @@ export async function getBufferFrom(
     cancellationToken?: CancellationToken): Promise<Buffer | null> {
 
     const streamEnded = new Deferred<void>();
-    while (size > 0) {
-        cancellationToken?.throwIfCancelled();
 
-        const readBuffer = readable.read(size) as Buffer;
-        if (readBuffer === null) {
-            const bytesAvailable = new Deferred<void>();
-            readable.once("readable", bytesAvailable.resolve.bind(bytesAvailable));
-            readable.once("end", streamEnded.resolve.bind(streamEnded));
-            const endPromise = Promise.race([bytesAvailable.promise, streamEnded.promise]);
-            await (cancellationToken ? cancellationToken.racePromise(endPromise) : endPromise);
+    if (size === 0) {
+        return Buffer.from([]);
+    } else if (size > 67108864) {
+        throw new Error("Package size excesses the limitation.");
+    }
 
-            if (bytesAvailable.isCompleted) {
-                continue;
+    let readBuffer: Buffer | null = null;
+    let index: number = 0;
+    while (size > 0) {
+        cancellationToken?.throwIfCancelled();
+        let availableSize = (readable as Readable).readableLength ?? size;
```

Comment:
I think you're treating the TypeScript `as` keyword as if it behaves the way C#'s does. In fact, it does not. `as` is a blind cast that always succeeds. It's only there to satisfy Typescript. The transpiled js for this line is:
```js
let availableSize = (_a = readable.readableLength) !== null && _a !== void 0 ? _a : size;
```

If `readableLength` is not defined on the object, this will be `undefined` rather than `null` and `availableSize` will get the `undefined` value.
You *can* do a type check for a class using [the expression `readable instanceof Readable`](https://www.tektutorialshub.com/typescript/typescript-instanceof-type-guard/), but in my experience that often produces `false` unexpectedly. The better approach I've found is to just access the property and check for `undefined` or the type you expect, and fallback if `undefined`.

The irony is that `??` in javascript works. But typescript transpiles it to that `!== null`, which breaks it. But you *can* use `||`, which TS leaves alone, and in javascript this works the way you want.